sysdig/resource_sysdig_secure_notification_channel_msteams.go

```diff
@@ -0,0 +1,213 @@
+package sysdig
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	v2 "github.com/draios/terraform-provider-sysdig/sysdig/internal/client/v2"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceSysdigSecureNotificationChannelMSTeams() *schema.Resource {
+	timeout := 5 * time.Minute
+
+	return &schema.Resource{
+		CreateContext: resourceSysdigSecureNotificationChannelMSTeamsCreate,
+		UpdateContext: resourceSysdigSecureNotificationChannelMSTeamsUpdate,
+		ReadContext:   resourceSysdigSecureNotificationChannelMSTeamsRead,
+		DeleteContext: resourceSysdigSecureNotificationChannelMSTeamsDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(timeout),
+			Update: schema.DefaultTimeout(timeout),
+			Read:   schema.DefaultTimeout(timeout),
+			Delete: schema.DefaultTimeout(timeout),
+		},
+
+		Schema: createSecureNotificationChannelSchema(map[string]*schema.Schema{
+			"url": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"template_version": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		}),
+	}
+}
+
+func resourceSysdigSecureNotificationChannelMSTeamsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := getSecureNotificationChannelClient(meta.(SysdigClients))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	teamID, err := client.CurrentTeamID(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	notificationChannel, err := secureNotificationChannelMSTeamsFromResourceData(d, teamID)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	notificationChannel, err = client.CreateNotificationChannel(ctx, notificationChannel)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(strconv.Itoa(notificationChannel.ID))
+	_ = d.Set("version", notificationChannel.Version)
+
+	return nil
+}
+
+func resourceSysdigSecureNotificationChannelMSTeamsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := getSecureNotificationChannelClient(meta.(SysdigClients))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id, _ := strconv.Atoi(d.Id())
+	nc, err := client.GetNotificationChannelById(ctx, id)
+
+	if err != nil {
+		d.SetId("")
```

Comment:
So if there is internal server error we are removing resource from state? Why are we removing resource from state when there is an error from API?

Reply:
I used the logic from another file, but I think the logic is to remove state from terraform so that next time, it can reconcile with infrastructure and get the correct state again. I will keep this as is for now for consistency

Reply:
Im not sure this is the logic we want to add because If I try to create a resource, and then read it and if that fails it will remove my resource from state but it will be on API. So I have to import it in order to use it but user will see no warnings nor errors on output. How would user know if he must import something? On the next terraform plan, tool will  probably suggest creating a resource which already exists and that might fail because notification channel with such name already exists...

Reply:
Understood now, I can return the err to the user

Reply:
Error should be returned but state should not be empty, remove also `d.SetId("")`

Reply:
We are doing it in all current monitor notification channel files as well. Might be good to consult a TF expert on this and remove it everywhere in a separate PR